tao_compiler/mlir/disc/transforms/lower_tf.cc

```diff
@@ -1080,6 +1083,161 @@ LogicalResult convertQuantizedConv2DWithBiasAndRequantizeOp(func::FuncOp func) {
+                                        builder->getIntegerType(64));
+  return mlir::DenseIntElementsAttr::get(
+      ty, llvm::makeArrayRef(values.data(), values.size()));
+}
+
+// Gets all float values from the given attribute and push them to `values`.
+static DenseFPElementsAttr GetF32ElementsAttr(Attribute attr, Builder *builder) {
+  SmallVector<float, 4> values;
+  auto array_attr = attr.cast<ArrayAttr>();
+  values.reserve(array_attr.getValue().size());
+  for (Attribute val : array_attr.getValue()) {
+    values.push_back(static_cast<float>(val.cast<FloatAttr>().getValueAsDouble()));
+  }
+  RankedTensorType ty = RankedTensorType::get(
+      {static_cast<int64_t>(values.size())}, builder->getF32Type());
+  return DenseFPElementsAttr::get(ty, values);
+}
+
+// Returns a 1-d i64 elements attribute populated with numbers from start to
+// end, excluding.
+static DenseIntElementsAttr GetI64ElementsAttrForSeq(int start, int end,
+                                                     Builder *builder) {
+  int size = end - start;
+
+  SmallVector<int64_t, 4> vals;
+  vals.resize(size);
+  std::iota(vals.begin(), vals.end(), start);
+
+  TensorType ty = RankedTensorType::get({size}, builder->getIntegerType(64));
+  return DenseIntElementsAttr::get(ty, vals);
+}
+
+// op {
+//   name: "Bucketize"
+//   input_arg {
+//     name: "input"
+//     type_attr: "T"
+//   }
+//   output_arg {
+//     name: "output"
+//     type: DT_INT32
+//   }
+//   attr {
+//     name: "T"
+//     type: "type"
+//     allowed_values {
+//       list {
+//         type: DT_INT32
+//         type: DT_INT64
+//         type: DT_FLOAT
+//         type: DT_DOUBLE
+//       }
+//     }
+//   }
+//   attr {
+//     name: "boundaries"
+//     type: "list(float)"
+//   }
+// }
+class ConvertBucketizeOp: public OpRewritePattern<TF::BucketizeOp> {
+ public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(TF::BucketizeOp op,
+                                PatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    Value input = op.input();
+    auto input_type = input.getType().dyn_cast<RankedTensorType>();
+    // attr: boundaries, type is float according op definition
+    auto boundaries = rewriter.create<mhlo::ConstOp>(
+            loc, GetF32ElementsAttr(op.boundaries(), &rewriter)
+        ).getResult();
+    // the following behavior matches the behavior of the core
+    // Bucketize kernel. However, comparing an int32 or int64 against float may
+    // lead to inaccurate bucketing due to rounding.
+    if (input_type.isF64()) {
+      boundaries = rewriter.create<mhlo::ConvertOp>(loc, boundaries, rewriter.getF64Type());
+    } else {
+      input = rewriter.create<mhlo::ConvertOp>(loc, input, rewriter.getF32Type());
+    }
+
+    // get input type again since it maybe converted
+    input_type = input.getType().dyn_cast<RankedTensorType>();
```

Comment:
nit: just  move this line to `else` above?

Reply:
done